File: pmod_swt_8ld/pmod_swt_8ld.py
import wiringpi
import sys

# Définition des pins des interrupteurs et des LEDs
switch_pins = [8, 10, 9, 11, 19, 21, 20, 18]
led_pins = [16, 14, 15, 17, 4, 12, 5, 6]

def read_switches():
    """Lit l'état des interrupteurs et retourne la valeur en hexadécimal."""
    switch_state = 0
    
    for i, pin in enumerate(switch_pins):
        if wiringpi.digitalRead(pin) == 1:  # HIGH
            switch_state |= (1 << i)
    
    print(f"État des interrupteurs : 0x{switch_state:X}")
    return switch_state

def control_leds(hex_value):
    """Met à jour l'état des LEDs en fonction de la valeur hexadécimale fournie."""
    for i, pin in enumerate(led_pins):
        state = (hex_value >> i) & 1
        wiringpi.digitalWrite(pin, state)

def main():
    port = sys.argv[1] if len(sys.argv) > 1 else "Inconnu"
    if port == "JA":
        print("Port JA sélectionné !")
    else:
        print(f"Port inconnu : {port}")

    if wiringpi.wiringPiSetupGpio() == -1:
        print("Erreur : impossible d'initialiser WiringPi")
        sys.exit(1)

    for pin in switch_pins:
        wiringpi.pinMode(pin, 0)
        wiringpi.pullUpDnControl(pin, 0)

    for pin in led_pins:
        wiringpi.pinMode(pin, 1)

    switch_state = read_switches()
    control_leds(switch_state)

if __name__ == "__main__":
    main()

File: pmod_swt_8ld/pmod_swt_8ld.c
#include <stdio.h>
#include <wiringPi.h>
#include <string.h>

#define SWITCH_COUNT 8  // Nombre d'interrupteurs
#define LED_COUNT 8     // Nombre de LED

int switch_pins[SWITCH_COUNT] = {8, 10, 9, 11, 19, 21, 20, 18};
int led_pins[LED_COUNT] = {16, 14, 15, 17, 4, 12, 5, 6};

int read_switches(int argc, char *argv[]) {
    int switch_state = 0;

    if (argc > 1 && strcmp(argv[1], "JA") == 0) {
        printf("Port JA sélectionné !\n");
    } else {
        printf("Port inconnu : %s\n", argv[1]);
    }

    // Initialisation de WiringPi
    if (wiringPiSetupGpio() == -1) { // Utilisation de la numérotation BCM
        perror("Erreur : impossible d'initialiser WiringPi");
        return 1;
    }

    // Configuration des broches en entrée
    for (int i = 0; i < SWITCH_COUNT; i++) {
        pinMode(switch_pins[i], INPUT);
        pullUpDnControl(switch_pins[i], PUD_DOWN); // Activation de la résistance pull-down
    }

    // Lire l'état des interrupteurs
    for (int i = 0; i < SWITCH_COUNT; i++) {
        if (digitalRead(switch_pins[i]) == HIGH) {
            switch_state |= (1 << i);
        }
    }

    printf("État des interrupteurs : 0x%X\n", switch_state);
    
    return switch_state;
}

void control_leds(unsigned char hex_value) {
    // Initialisation des broches des LEDs
    for (int i = 0; i < LED_COUNT; i++) {
        pinMode(led_pins[i], OUTPUT);
    }

    // Mise à jour des LEDs selon la valeur hexadécimale
    for (int i = 0; i < LED_COUNT; i++) {
        int state = (hex_value >> i) & 1;  // Récupération du bit correspondant
        digitalWrite(led_pins[i], state);
    }
}

int main(int argc, char *argv[]) {
    wiringPiSetup();  // Initialisation de la bibliothèque WiringPi
    unsigned char hex_value = read_switches(argc, argv);
    control_leds(hex_value);
    return 0;
}
